src/TopDownProteomics/ProForma/ChemicalProteoformHashGenerator.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using TopDownProteomics.Chemistry;
+using TopDownProteomics.Proteomics;
+
+namespace TopDownProteomics.ProForma
+{
+    /// <summary>Creates a chemical proteoform hash from a proteoform group.</summary>
+    public class ChemicalProteoformHashGenerator
+    {
+        /// <summary>Generates a chemical proteoform hash for the specified proteoform group.</summary>
+        /// <param name="proteoformGroup">The proteoform group.</param>
+        /// <returns></returns>
+        public string Generate(IProteoformGroup proteoformGroup)
+        {
+            ProFormaDescriptor nTermDescriptor =
+                this.GetFormulaDescriptor(proteoformGroup.NTerminalModification);
+            IList<ProFormaDescriptor> nTermDescriptors = nTermDescriptor == null ? null : new[] { nTermDescriptor };
+
+            ProFormaDescriptor cTermDescriptor =
+                this.GetFormulaDescriptor(proteoformGroup.CTerminalModification);
+            IList<ProFormaDescriptor> cTermDescriptors = cTermDescriptor == null ? null : new[] { cTermDescriptor };
+
+            IList<ProFormaTag> tags = new List<ProFormaTag>();
+            if (proteoformGroup.Modifications != null)
```

Comment:
I'd check length here and only create a tags List<> as needed, i.e. 
`if (proteoformGroup.Modifications?.Length > 0)
{
tags = new List<ProFormaTag>();
...}`

Reply:
I think my code was wrong, should be more like:
`if ((proteoformGroup.Modifications?.Length ?? 0) > 0) { tags = new List<ProFormaTag>(); ...}`
Kind of ugly ...